File: 2830_maximize_the_profit_as_the_salesman/Solution.java
// LeetCode 2830 - Maximize the Profit as the Salesman
// https://leetcode.com/problems/maximize-the-profit-as-the-salesman/

import java.util.ArrayList;
import java.util.List;

class Solution {
    public int maximizeTheProfit(int n, List<List<Integer>> offers) {
        var byEnd = new ArrayList<List<Integer>>[n];
        for (int i = 0; i < n; i++) byEnd.set(i, new ArrayList<List<Integer>>());
        for (var o : offers) byEnd.get(o[1]).add(o);
        int[] dp = new int[n + 1];
        for (int end = 0; end < n; end++) {
            dp[end + 1] = dp[end];
            for (var o : byEnd.get(end))
                dp[end + 1] = Math.max(dp[end + 1], dp[o[0]] + o[2]);
        }
        return dp[n];
    }
}

File: 2830_maximize_the_profit_as_the_salesman/solution.cpp
// LeetCode 2830 - Maximize the Profit as the Salesman
// https://leetcode.com/problems/maximize-the-profit-as-the-salesman/

#include <algorithm>
#include <vector>

class Solution {
public:
    int maximizeTheProfit(int n, std::vector<std::vector<int>>& offers) {
        std::vector<std::vector<std::vector<int>>> byEnd(n);
        for (auto& o : offers) byEnd[o[1]].push_back(o);
        std::vector<int> dp(n + 1, 0);
        for (int end = 0; end < n; end++) {
            dp[end + 1] = dp[end];
            for (auto& o : byEnd[end]) {
                dp[end + 1] = std::max(dp[end + 1], dp[o[0]] + o[2]);
            }
        }
        return dp[n];
    }
};
